server/vulnerabilities/nvd/sync/cve_syncer.go

```diff
@@ -0,0 +1,756 @@
+// Package nvdsync provides a CVE syncer that uses the NVD 2.0 API to download CVE information
+// and stores it in the legacy format. The reason we decided to store in the legacy format is because
+// the github.com/facebookincubator/nvdtools doesn't yet support parsing the new API 2.0 JSON format.
+package nvdsync
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/fs"
+	"net/http"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"time"
+
+	"github.com/facebookincubator/nvdtools/cvefeed/nvd/schema"
+	"github.com/fleetdm/fleet/v4/orbit/pkg/constant"
+	"github.com/fleetdm/fleet/v4/pkg/fleethttp"
+	"github.com/fleetdm/fleet/v4/server/ptr"
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/pandatix/nvdapi/common"
+	"github.com/pandatix/nvdapi/v2"
+)
+
+// CVE syncs CVE information from the NVD database (nvd.nist.gov) using its API 2.0.
+// It stores the CVE information using the legacy feed format.
+// The reason we decided to store in the legacy format is because
+// the github.com/facebookincubator/nvdtools doesn't yet support parsing
+// the new API 2.0 JSON format.
+type CVE struct {
+	client *http.Client
+	dbDir  string
+	logger log.Logger
+	debug  bool
+}
+
+var (
+	// timeBetweenRequests is the recommended time to wait between NVD API requests.
+	timeBetweenRequests = 6 * time.Second
+	// maxRetryAttempts is the maximum number of request to retry in case of API failure.
+	maxRetryAttempts = 10
+	// waitTimeForRetry is the time to wait between retries.
+	waitTimeForRetry = 30 * time.Second
+)
+
+// CVEOption allows configuring a CVE syncer.
+type CVEOption func(*CVE)
+
+// WithLogger sets the logger for a CVE syncer.
+//
+// Default value is log.NewNopLogger().
+func WithLogger(logger log.Logger) CVEOption {
+	return func(s *CVE) {
+		s.logger = logger
+	}
+}
+
+// WithDebug sets the debug mode for a CVE syncer.
+//
+// Default value is false.
+func WithDebug(debug bool) CVEOption {
+	return func(s *CVE) {
+		s.debug = debug
+	}
+}
+
+// NewCVE creates and returns a CVE syncer.
+// The provided dbDir is the local directory to use to store/update
+// CVE information from NVD.
+func NewCVE(dbDir string, opts ...CVEOption) (*CVE, error) {
+	if dbDir == "" {
+		return nil, errors.New("directory not set")
+	}
+	s := CVE{
+		client: fleethttp.NewClient(),
+		dbDir:  dbDir,
+		logger: log.NewNopLogger(),
+	}
+	for _, fn := range opts {
+		fn(&s)
+	}
+	return &s, nil
+}
+
+func (s *CVE) lastModStartDateFilePath() string {
+	return filepath.Join(s.dbDir, "last_mod_start_date.txt")
+}
+
+// Do runs the synchronization from the NVD service to the local DB directory.
+func (s *CVE) Do(ctx context.Context) error {
+	ok, err := fileExists(s.lastModStartDateFilePath())
+	if err != nil {
+		return err
+	}
+	if !ok {
+		level.Debug(s.logger).Log("msg", "initial NVD CVE sync")
+		return s.initSync(ctx)
+	}
+	level.Debug(s.logger).Log("msg", "NVD CVE update")
+	return s.update(ctx)
```

Comment:
Should we fallback to doing an initial sync if update returns error?

Reply:
Agreed. But is there a way for us or customer to be alerted that CVE updates are not happening? The updates can be broken if they have a corrupted file, right?